File: acme-support-agent/typescript/src/agent.ts
/**
 * The Acme agent — instrumented with the OpenSearch GenAI SDK for TypeScript.
 *
 * observe({ op: Op.INVOKE_AGENT }) wraps the whole turn as the root span; the
 * tools (wrapped in tools.ts) emit execute_tool spans; the chat call is wrapped
 * with observe({ op: Op.CHAT }) and enriched with token usage. This is the direct
 * TS equivalent of the Python register()+observe()+enrich() flow.
 */

import OpenAI from "openai";
import { observe, enrich, Op } from "@opensearch-project/genai-observability-sdk-ts";
import { TOOL_FUNCTIONS, TOOL_SCHEMAS, SYSTEM_PROMPT } from "./tools.js";

const MODEL = process.env.ACME_MODEL ?? "gpt-4o";

function openaiTools() {
  return TOOL_SCHEMAS.map((s) => ({ type: "function" as const, function: s }));
}

// One chat turn against the model, wrapped as a `chat` span with token usage.
const chatStep = observe(
  { name: "chat", op: Op.CHAT },
  async function chat(client: OpenAI, messages: any[]) {
    const resp = await client.chat.completions.create({
      model: MODEL,
      messages,
      tools: openaiTools(),
    });
    const usage = resp.usage;
    if (usage) {
      enrich({
        model: MODEL,
        provider: "openai",
        inputTokens: usage.prompt_tokens,
        outputTokens: usage.completion_tokens,
        totalTokens: usage.total_tokens,
      });
    }
    return resp.choices[0].message;
  },
);

export const handleSupportQuestion = observe(
  { name: "acme-support-agent", op: Op.INVOKE_AGENT },
  async function handleSupportQuestion(
    question: string,
    conversationId = "ts-session",
  ): Promise<string> {
    enrich({ model: MODEL, provider: "openai", sessionId: conversationId });

    const client = new OpenAI();
    const messages: any[] = [
      { role: "system", content: SYSTEM_PROMPT },
      { role: "user", content: question },
    ];

    for (let step = 0; step < 5; step++) {
      const msg = await chatStep(client, messages);

      if (!msg.tool_calls?.length) {
        return msg.content ?? "";
      }

      messages.push(msg);
      for (const call of msg.tool_calls) {
        const fn = TOOL_FUNCTIONS[call.function.name];
        const args = JSON.parse(call.function.arguments || "{}");
        const result = fn(args); // execute_tool span emitted by the wrapped tool
        messages.push({
          role: "tool",
          tool_call_id: call.id,
          content: JSON.stringify(result),
        });
      }
    }

    return "Sorry, I couldn't complete that request.";
  },
);

File: acme-support-agent-agentcore/agent.py
"""Acme Support Agent on Amazon Bedrock AgentCore Runtime — one clean flow.

AgentCore Runtime hosts the agent behind a managed HTTP endpoint. The contract is
a single @app.entrypoint function that takes the request payload and returns the
response. Everything else — the agent loop, the tools, the observability — is the
same shared core used by every other variant.

We instrument inside the entrypoint so each invocation is one invoke_agent trace,
exported to the OpenSearch stack via OTLP. In AgentCore, point
OTEL_EXPORTER_OTLP_TRACES_ENDPOINT at your collector (in-VPC for production).

Local dev:    python agent.py            # runs the AgentCore dev server on :8080
Deploy:       see README.md (agentcore configure && agentcore launch)

Docs: https://aws.amazon.com/bedrock/agentcore/
"""

from __future__ import annotations

from bedrock_agentcore import BedrockAgentCoreApp

from acme_shared import setup_observability, observe, enrich, Op
from acme_shared.tools import lookup_order, check_inventory, search_policy, SYSTEM_PROMPT

# Configure observability once at import/cold-start.
setup_observability(service_name="acme-support-agent")

app = BedrockAgentCoreApp()

# Default model is a Bedrock-hosted Claude; AgentCore runs with the task role's
# AWS credentials, so no keys are needed in the container.
MODEL = "anthropic.claude-3-5-sonnet-20240620-v1:0"


def _bedrock_tool_config():
    from acme_shared.tools import TOOL_SCHEMAS
    return {"tools": [
        {"toolSpec": {"name": s["name"], "description": s["description"],
                      "inputSchema": {"json": s["parameters"]}}}
        for s in TOOL_SCHEMAS
    ]}


_TOOLS = {"lookup_order": lookup_order, "check_inventory": check_inventory,
          "search_policy": search_policy}


@observe(op=Op.INVOKE_AGENT, name="acme-support-agent")
def _run_agent(question: str, conversation_id: str) -> str:
    import boto3

    enrich(provider="aws.bedrock", model=MODEL,
           session_id=conversation_id)  # session_id -> gen_ai.conversation.id
    client = boto3.client("bedrock-runtime")
    messages = [{"role": "user", "content": [{"text": question}]}]

    for _ in range(5):
        resp = client.converse(
            modelId=MODEL, messages=messages,
            system=[{"text": SYSTEM_PROMPT}], toolConfig=_bedrock_tool_config(),
        )
        out = resp["output"]["message"]
        messages.append(out)
        if resp.get("stopReason") != "tool_use":
            return "".join(b["text"] for b in out["content"] if "text" in b)
        results = []
        for block in out["content"]:
            if "toolUse" in block:
                tu = block["toolUse"]
                r = _TOOLS[tu["name"]](**tu["input"])
                results.append({"toolResult": {"toolUseId": tu["toolUseId"],
                                                "content": [{"json": r}]}})
        messages.append({"role": "user", "content": results})
    return "Sorry, I couldn't complete that request."


@app.entrypoint
def invoke(payload: dict) -> dict:
    """AgentCore Runtime entrypoint.

    payload: {"prompt": "...", "conversation_id": "..."} (conversation_id optional)
    returns: {"response": "..."}
    """
    question = payload.get("prompt", "")
    conversation_id = payload.get("conversation_id", "agentcore-session")
    answer = _run_agent(question, conversation_id)
    return {"response": answer}


if __name__ == "__main__":
    # Local AgentCore dev server (mirrors the deployed runtime contract).
    app.run()
